kube-core/src/version.rs

```diff
@@ -1,18 +1,62 @@
-//! Version definitions to allow sorting (not exported)
-//!
-//! Follows [kubernetes version priority](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority)
-//! but only in sort-order at the moment.
-//!
-//! Changes to Ord is needed if this is ever exported outside the crate.
+use std::{cmp::Reverse, convert::Infallible, str::FromStr};
 
-use std::cmp::Reverse;
-
-#[derive(PartialEq, Eq, Debug)]
-pub(crate) enum Version {
+/// Version parser for Kubernetes version patterns
+///
+/// Follows [Kubernetes version priority](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority)
+/// to allow getting the correct sort-order:
+///
+/// ```
+/// use kube_core::Version;
+/// use std::cmp::Reverse; // for DESCENDING sort
+/// let mut versions = vec![
+///     "v10beta3",
+///     "v2",
+///     "foo10",
+///     "v1",
+///     "v3beta1",
+///     "v11alpha2",
+///     "v11beta2",
+///     "v12alpha1",
+///     "foo1",
+///     "v10",
+/// ];
+/// versions.sort_by_cached_key(|v| Reverse(Version::parse(v)));
+/// assert_eq!(versions, vec![
+///     "v10",
+///     "v2",
+///     "v1",
+///     "v11beta2",
+///     "v10beta3",
+///     "v3beta1",
+///     "v12alpha1",
+///     "v11alpha2",
+///     "foo1",
+///     "foo10",
+/// ]);
+/// ```
+///
+/// and corect direct comparisons after parsing:
+///
+/// ```
+/// use kube_core::Version;
+/// assert!(Version::Stable(2) > Version::Stable(1));
+/// assert!(Version::Stable(1) > Version::Beta(1, None));
+/// assert!(Version::Stable(1) > Version::Beta(2, None));
+/// assert!(Version::Stable(2) > Version::Alpha(1, Some(2)));
+/// assert!(Version::Stable(1) > Version::Alpha(2, Some(2)));
+/// assert!(Version::Beta(1, None) > Version::Nonconformant("ver3".into()));
+/// ```
+#[derive(PartialEq, Eq, Debug, Clone)]
+pub enum Version {
+    /// A major/GA release
     Stable(u32),
+    /// A beta release for a specific major version
     Beta(u32, Option<u32>),
+    /// An alpha release for a specific major version
     Alpha(u32, Option<u32>),
-    // CRDs and APIServices can use arbitrary strings as versions.
+    /// An non-conformant api string (sorted lexicographically)
```

Comment:
This isn't really new, but this might be worth reconsidering before stabilizing. Keeping this "everything else" case means that anyone relying on it will silently be broken if we add new primary cases.

Instead, we might want to turn `Nonconformant` into the error case on `<Version as FromStr>`?

Reply:
According to [Version Priority](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definition-versioning/#version-priority) any non-conformant value is legal. I may want to set my CRD version to `testing` and expect it to work with k8s. So failing the parser on that will make it unusable from `kube`

Reply:
To be clear: I am not saying that a non-conformant version is illegal, or that we should reject working with it. Rather, that we should treat it as an opaque `String` in general, parse it into a `Version` when needed (such as for comparisons), and accept that that parsing may fail (with reasonable fallback behaviour, as described in the PR body).